corehq/apps/commtrack/processing.py

```diff
@@ -154,3 +173,54 @@ def process_stock(xform, case_db=None):
         relevant_cases=relevant_cases,
         stock_report_helpers=stock_report_helpers,
     )
+
+
+def _adjust_ledger_values(original_balance, stock_transaction):
+    if stock_transaction.report.type == stockconst.REPORT_TYPE_BALANCE:
+        quantity = stock_transaction.stock_on_hand
+    elif stock_transaction.report.type == stockconst.REPORT_TYPE_TRANSFER:
+        quantity = stock_transaction.quantity
+    else:
+        raise ValueError()
+
+    ledger_values = compute_ledger_values(
+        lambda: original_balance, stock_transaction.report.type, quantity)
+
+    if stock_transaction.report.type == stockconst.REPORT_TYPE_BALANCE:
+        assert stock_transaction.stock_on_hand == ledger_values.balance
+        # this is currently always 0 for some reason
+        stock_transaction.quantity = ledger_values.delta
+    elif stock_transaction.report.type == stockconst.REPORT_TYPE_TRANSFER:
+        stock_transaction.stock_on_hand = ledger_values.balance
+        assert stock_transaction.quantity == ledger_values.delta, (stock_transaction.quantity, ledger_values.delta)
+    else:
+        raise ValueError()
+
+    return ledger_values.balance
+
+
+@transaction.atomic
+def rebuild_stock_state(case_id, section_id, product_id):
+    """
+    rebuilds the StockState object
+    and the quantity and stock_on_hand fields of StockTransaction
+    when they are calculated from previous state
+    (as opposed to part of the explict transaction)
+
+    """
+
+    # these come out latest first, so reverse them below
+    stock_transactions = (
+        StockTransaction
+        .get_ordered_transactions_for_stock(
+            case_id=case_id, section_id=section_id, product_id=product_id)
+        .select_related('report__type')
+    )
+    balance = None
+    for stock_transaction in reversed(stock_transactions):
+        if stock_transaction.subtype == stockconst.TRANSACTION_SUBTYPE_INFERRED:
+            print 'deleting stock transaction'
+            stock_transaction.delete()
+        else:
+            balance = _adjust_ledger_values(balance, stock_transaction)
+            stock_transaction.save()
```

Comment:
This is the actual rebuilding. This function is so short mostly because
- `_adjust_ledger_values` takes care of the calculations
- all the stock state adjustment is done through the signals

if/when we refactor stock transactions not to use signals, we'll have to make sure the right things get called after this to rebuild stock state.